eqt/ui/UIFormWidget.py

```diff
@@ -41,58 +41,199 @@ def createForm(self):
+                if value >= widget_number:
+                    self.widget_number_dictionary[key] = value + 1
+            self.widget_number_dictionary[name] = widget_number
+
+    def _popWidgetNumberDictionary(self, name):
+        '''
+        Removes one item in the widget-number dictionary whose key is name.
+        As one widget is removed, the widget numbers associated to the other widgets 
+        in the layout are updated.
+
+        Parameters:
+        ---------------
+        name: string
+            name of the widget
+        '''
+        widget_number = self.getWidgetNumber(name)
+        for key, value in self.widget_number_dictionary.items():
+            if value > widget_number:
+                self.widget_number_dictionary[key] = value - 1
+        self.widget_number_dictionary.pop(name) 
 
+    def _popWidgetFromDictionary(self, dictionary, name):
+        '''
+        Removes the item(s) associated with `name` from a dictionary.
+
+        Parameters:
+        -----------------
+        dictionary :  dict
+        name: str
+            Format: {name}
+        '''
+        if name in dictionary.keys():
+            dictionary.pop(name) 
+        if f'{name}_field' in dictionary.keys():
+            dictionary.pop(f'{name}_field')      
+        if f'{name}_label' in dictionary.keys():
+            dictionary.pop(f'{name}_label') 
+
+    def addWidget(self, qwidget, qlabel, name):
+        '''
+        Adds a widget and a label widget, or a spanning widget, at the the end of
+        the form layout. 
+
+        Parameters:
+        ----------
+        name: str
+        qwidget: widget
+        qlabel: qlabel widget or str
+        '''
+        self._addWidget(name, qwidget, qlabel)
+    
+    def addSpanningWidget(self, qwidget, name):
+        '''Adds a spanning qwidget occupying the full row in the form layout.'''
+        self._addWidget(name, qwidget)
+        
     def getNumWidgets(self):
         '''
         Returns the number of widgets in the form.
         '''
         return self.num_widgets
 
-    def getWidget(self, name, role='field'):
-        '''returns the Widget by the name with which it has been added
+    def removeWidget(self, name):
+        '''
+        If not present already, creates a dictionary to store the removed qwidgets.
+        Sets the parent of the qwidget, and qlabel if present, to `None` and 
+        stores the widgets in the removed-widgets dictionary.
+        Deletes the row in the form layout.
+        Deletes the qwidget and qlabel from the widgets dictionary.
+        Deletes the widget number from the widget-number dictionary.
+    
+        Parameters:
+        --------------
+        name : str
+            name of the widget to be removed
+        '''
+        formLayout = self.uiElements['groupBoxFormLayout']
+        if not hasattr(self, 'removed_widgets_dictionary'):
+            self.removed_widgets_dictionary = {}
```

Comment:
I would define this in the `__init__` so that we know this dictionary exists, although empty.
I don't quite like to use `hasattr`